### PR TITLE
[runtime] Convert a lot of hashtables to be concurrent.

### PR DESCRIPTION
**Pull request by kumpera**
This significantly reduces Roslyn contention when JIT'ing.

### CODE REVIEW

File: mono/metadata/loader.c [hunk 1]
```diff
@@ -1727,6 +1727,16 @@ mono_get_method_full (MonoImage *image, guint32 token, MonoClass *klass,
 	return result;
 }
 
+static void
+create_conc_hash_table_if_needed (MonoConcurrentHashTable **ptr)
+{
+	MonoConcurrentHashTable *val = mono_conc_hashtable_new (NULL, NULL);
```
Comment (kumpera) [inline mono/metadata/loader.c]: Fixed

File: mono/metadata/metadata.c [hunk 2]
```diff
@@ -3019,25 +3019,30 @@ mono_metadata_get_generic_inst (int type_argc, MonoType **type_argv)
...
 
+	mono_image_set_lock (set);
+	MonoGenericInst *ginst2 = (MonoGenericInst *)mono_conc_hashtable_insert (set->ginst_cache, ginst, ginst);
 	mono_image_set_unlock (set);
-	return ginst;
+
+	if (!ginst2)
+		 ginst2 = ginst;
+	else
+		free_generic_inst (ginst);
```
Comment (vargaz) [inline mono/metadata/metadata.c]: Didn't the old value get overwritten in this case ?

Comment (vargaz) [inline mono/metadata/metadata.c]: Can this be added as an api comment ? Normal hashtable inserts replace the old values.

Comment (kumpera) [inline mono/metadata/metadata.c]: Will clarify the existing one.

**Comment by kumpera** [comment]
cc @vkargov

**Comment by vargaz** [comment]
Looks ok.

**Comment by kumpera** [comment]
Rebased on top of master from right now. cc @vargaz @marek-safar

**Comment by marek-safar** [comment]
There looks to be something wrong in your changes. The invalid IL errors are popping up on random places even if IL code is always same and coming from external binary. It looks like some race in runtime again.

**Comment by kumpera** [comment]
uggg, I'll do 9 PRs then to see if it helps pinpointing the bad one

**Comment by vargaz** [comment]
Its possible that this pr is good, but it just exposes a race elsewhere.

**Comment by marek-safar** [comment]
The crashes happens in same types we had problems before Zoltan put big load lock but it could be just coincidence

**Pull request closed without merging**